test/e2e/performanceprofile/functests/13_llc/llc.go

```diff
@@ -239,3 +357,137 @@ func addContent(ignitionConfig *igntypes.Config, content []byte, dst string, mod
 		},
 	})
 }
+
+func MachineFromJSON(data string) (Machine, error) {
+	ma := Machine{}
+	rd := strings.NewReader(data)
+	err := json.NewDecoder(rd).Decode(&ma)
+	return ma, err
+}
+
+func isCPUSetLLCAligned(infos []CacheInfo, cpus cpuset.CPUSet) (bool, *CacheInfo) {
+	for idx := range infos {
+		info := &infos[idx]
+		if cpus.IsSubsetOf(info.CPUs) {
+			return true, info
+		}
+	}
+	return false, nil
+}
+
+func computeLLCLayout(mi Machine) []CacheInfo {
+	ret := []CacheInfo{}
+	for _, node := range mi.Topology.Nodes {
+		for _, cache := range node.Caches {
+			if cache.Level < 3 { // TODO
+				continue
+			}
+			ret = append(ret, CacheInfo{
+				NodeID: node.ID,
+				Level:  int(cache.Level),
+				CPUs:   cpusetFromLogicalProcessors(cache.LogicalProcessors...),
+			})
+		}
+	}
+	return ret
+}
+
+func cpusetFromLogicalProcessors(procs ...uint32) cpuset.CPUSet {
+	cpuList := make([]int, 0, len(procs))
+	for _, proc := range procs {
+		cpuList = append(cpuList, int(proc))
+	}
+	return cpuset.New(cpuList...)
+}
+
+func expectedL3GroupSize(md MachineData) int {
+	// TODO: we assume all L3 Groups are equal in size.
+	for idx := range md.Caches {
+		cache := &md.Caches[idx]
+		if cache.Level != 3 {
+			continue
+		}
+		return cache.CPUs.Size()
+	}
+	return 0
+}
+
+func collectMachineDatas(ctx context.Context, nodeList []corev1.Node) (map[string]MachineData, error) {
+	cmd := []string{"/usr/bin/machineinfo"}
+	infos := make(map[string]MachineData)
+	for idx := range nodeList {
+		node := &nodeList[idx]
+		out, err := nodes.ExecCommand(ctx, node, cmd)
+		if err != nil {
+			return infos, err
+		}
+
+		info, err := MachineFromJSON(string(out))
```

Comment:
why do we need to convert this to string ? 
can we do this directly ?
```
ma = Machine {}
err := json.Unmarshal(out, &ma)
```

Reply:
I can use a bytes.Buffer indeed